File: src/test/java/com/ealanta/productapp/product/ProductComparatorTest.java
package com.ealanta.productapp.product;

import com.ealanta.productapp.item.PriceDetails;
import com.ealanta.productapp.misc.Constants;
import com.ealanta.productapp.price.PriceInfo;
import com.ealanta.productapp.price.PriceReduction;
import org.javamoney.moneta.Money;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;

import javax.money.MonetaryAmount;
import java.math.BigDecimal;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

public class ProductComparatorTest{

    @Before
    public void setup() {
        pr1 = getProduct("pr1", Optional.of("1.00"));
        pr2 = getProduct("pr2", Optional.of("2.00"));
        pr3 = getProduct("pr3", Optional.of("3.00"));

        nopr1 = getProduct("noPr1", Optional.empty());
        nopr2 = getProduct("noPr2", Optional.empty());
        nopr3 = getProduct("noPr3", Optional.empty());
    }

    private Product getProduct(String productId, Optional<String> reductionAmount) {
        Product result = new Product();
        result.setProductId(productId);
        PriceDetails details = new PriceDetails();
        details.setCurrency(Constants.DEFAULT_CURRENCY_CODE);
        PriceInfo info = new PriceInfo();
        Optional<PriceReduction> optPR = reductionAmount.map(ra -> new PriceReduction(getMoney(ra), getMoney("0.00")));
        info.setPriceReduction(optPR);
        return result;
    }

    Product pr1;
    Product pr2;
    Product pr3;

    Product nopr1;
    Product nopr2;
    Product nopr3;

    private ProductComparator comparator = new ProductComparator();

    private List<Product> sortAscending(List<Product> source) {
        List<Product> sorted = source.stream().sorted(comparator).collect(Collectors.toList());
        return sorted;
    }

    private List<Product> sortDescending(List<Product> source) {
        List<Product> sorted = source.stream().sorted(comparator.reversed()).collect(Collectors.toList());
        return sorted;
    }

    MonetaryAmount getMoney(String amt) {
        return Money.of(new BigDecimal(amt).setScale(2), "GBP");
    }

    @Test
    public void confirmMonetaryAmountComparator() {
        MonetaryAmount m1 = getMoney("10");
        MonetaryAmount m2 = getMoney("100");
        Assert.assertTrue(m1.compareTo(m2) < 0);
        Assert.assertTrue(m2.compareTo(m1) > 0);
        Assert.assertTrue(m2.compareTo(m2) == 0);
    }

    @Test
    public void testProductComparator1() {
        Assert.assertTrue(comparator.compare(pr1, pr2) < 0);
        Assert.assertTrue(comparator.compare(pr2, pr1) > 0);
        Assert.assertTrue(comparator.compare(pr2, pr2) == 0);
    }

    @Test
    public void testSortDescendingByPriceRedution() {
        List<Product> products = Arrays.asList(pr1, pr2, pr3);
        List<Product> sorted = sortDescending(products);
        Assert.assertEquals(Arrays.asList(pr3, pr2, pr1), sorted);
    }

    @Test
    public void testSortAscendingByPriceReduction() {
        List<Product> products = Arrays.asList(pr3, pr2, pr1);
        List<Product> sorted = sortAscending(products);
        Assert.assertEquals(Arrays.asList(pr1, pr2, pr3), sorted);
    }

    @Test
    public void testSortDescendingByNoPriceReduction() {
        List<Product> products = Arrays.asList(nopr1, nopr2, nopr3);
        List<Product> sorted = sortDescending(products);
        Assert.assertEquals(Arrays.asList(nopr3, nopr2, nopr1), sorted);
    }

    @Test
    public void testSortAscendingByNoPriceReduction() {
        List<Product> products = Arrays.asList(nopr3, nopr2, nopr1);
        List<Product> sorted = sortAscending(products);
        Assert.assertEquals(Arrays.asList(nopr1, nopr2, nopr3), sorted);
    }

    @Test
    public void testSortAscendingMixed() {
        List<Product> products = Arrays.asList(pr1, nopr1);
        List<Product> sorted = sortAscending(products);
        Assert.assertEquals(Arrays.asList(nopr1, pr1), sorted);
    }

    @Test
    public void testSortDescendingMixed() {
        List<Product> products = Arrays.asList(nopr1, pr1);
        List<Product> sorted = sortDescending(products);
        Assert.assertEquals(Arrays.asList(pr1, nopr1), sorted);
    }

}

File: src/test/kotlin/com/ealanta/productapp/product/ProductComparatorTest.kt
package com.ealanta.productapp.product

import com.ealanta.productapp.config.ProductComparatorConfig
import com.ealanta.productapp.domain.PriceDetails
import com.ealanta.productapp.misc.Constants
import com.ealanta.productapp.domain.PriceInfo
import com.ealanta.productapp.domain.PriceReduction
import com.ealanta.productapp.domain.Product
import org.javamoney.moneta.Money
import org.junit.Before
import org.junit.Test

import javax.money.MonetaryAmount
import java.math.BigDecimal
import java.util.Arrays
import kotlin.properties.Delegates
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class ProductComparatorTest {

    private var pr1: Product by Delegates.notNull()
    private var pr2: Product by Delegates.notNull()
    private var pr3: Product by Delegates.notNull()

    private var nopr1: Product by Delegates.notNull()
    private var nopr2: Product by Delegates.notNull()
    private var nopr3: Product by Delegates.notNull()

    private val comparator:Comparator<Product> = ProductComparatorConfig().productComparator()

    @Before
    fun setup() {
        pr1 = getProduct("pr1", "1.00")
        pr2 = getProduct("pr2", "2.00")
        pr3 = getProduct("pr3", "3.00")

        nopr1 = getProduct("noPr1", null)
        nopr2 = getProduct("noPr2", null)
        nopr3 = getProduct("noPr3", null)
    }

    private fun getProduct(productId: String, reductionAmount: String?): Product {
        val result = Product()
        result.productId = productId
        val details = PriceDetails()
        details.currency = Constants.DEFAULT_CURRENCY_CODE
        val info = PriceInfo()
        info.priceReduction = reductionAmount?.let { ra -> PriceReduction(getMoney(ra), getMoney("0.00")) }
        return result
    }

    private fun sortAscending(source: List<Product>): List<Product> {
        return source.sortedWith(comparator)
    }

    private fun sortDescending(source: List<Product>): List<Product> {
        return source.sortedWith(comparator.reversed())
    }

    internal fun getMoney(amt: String): MonetaryAmount {
        return Money.of(BigDecimal(amt).setScale(2), "GBP")
    }

    @Test
    fun confirmMonetaryAmountComparator() {
        val m1 = getMoney("10")
        val m2 = getMoney("100")
        assertTrue(m1.compareTo(m2) < 0)
        assertTrue(m2.compareTo(m1) > 0)
        assertTrue(m2.compareTo(m2) == 0)
    }

    @Test
    fun testProductComparator1() {
        assertTrue(comparator.compare(pr1, pr2) < 0)
        assertTrue(comparator.compare(pr2, pr1) > 0)
        assertTrue(comparator.compare(pr2, pr2) == 0)
    }

    @Test
    fun testSortDescendingByPriceRedution() {
        val products = Arrays.asList(pr1, pr2, pr3)
        val sorted = sortDescending(products)
        assertEquals(Arrays.asList(pr3, pr2, pr1), sorted)
    }

    @Test
    fun testSortAscendingByPriceReduction() {
        val products = Arrays.asList(pr3, pr2, pr1)
        val sorted = sortAscending(products)
        assertEquals(Arrays.asList(pr1, pr2, pr3), sorted)
    }

    @Test
    fun testSortDescendingByNoPriceReduction() {
        val products = Arrays.asList(nopr1, nopr2, nopr3)
        val sorted = sortDescending(products)
        assertEquals(Arrays.asList(nopr3, nopr2, nopr1), sorted)
    }

    @Test
    fun testSortAscendingByNoPriceReduction() {
        val products = Arrays.asList(nopr3, nopr2, nopr1)
        val sorted = sortAscending(products)
        assertEquals(Arrays.asList(nopr1, nopr2, nopr3), sorted)
    }

    @Test
    fun testSortAscendingMixed() {
        val products = Arrays.asList(pr1, nopr1)
        val sorted = sortAscending(products)
        assertEquals(Arrays.asList(nopr1, pr1), sorted)
    }

    @Test
    fun testSortDescendingMixed() {
        val products = Arrays.asList(nopr1, pr1)
        val sorted = sortDescending(products)
        assertEquals(Arrays.asList(pr1, nopr1), sorted)
    }

}
